strawberryfields/gbs/data.py

```diff
@@ -0,0 +1,199 @@
+[2, 0, 8, 11, ... , 6]
+
+Code details
+^^^^^^^^^^^^
+
+The :class:`Dataset` class provides the base functionality from which all datasets such as
+:class:`Planted` inherit.
+"""
+# pylint: disable=unnecessary-pass
+from abc import ABCMeta, abstractmethod
+
+import pkg_resources
+import numpy as np
+import scipy
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(metaclass=ABCMeta):
+    """Base class for loading datasets of pre-generated samples.
+
+    Attributes:
+        n_mean (float): mean number of photons in the GBS device
+        n_max (float): maximum number of photons allowed in any sample. This number is set to
+            limit the computation time, any sample being simulated that exceeds ``n_max`` will be
+            ignored and cause calculation to skip to the next sample.
+        threshold (bool): flag to indicate whether samples are generated with threshold detection
+            (i.e., detectors of zero or some photons) or with photon number resolving detectors.
+        adj (array): adjacency matrix of graph from which samples were generated
+        n_samples (int): total number of samples in dataset
+        modes (int): number of modes in GBS device or, equivalently, number of nodes in graph
+    """
+
+    _count = 0
+
+    @property
+    @abstractmethod
+    def _data_filename(self) -> str:
+        """Name of files stored in ``./data/``. Samples should be provided as a
+        ``scipy.sparse.csr_matrix`` saved in ``.npz`` format and the corresponding adjacency
+        matrix should be provided as a ``.npy`` binary. For ``_data_filename = "example"``,
+        the corresponding samples should be stored as ``./data/example.npz`` and the adjacency
+        matrix as ``./data/example_A.npy``."""
+        pass
+
+    def __init__(self):
+        """Instantiating ``Dataset`` causes the ``scipy.sparse.csr_matrix`` of samples and
+        ``numpy.ndarray`` adjacency matrix to be loaded."""
+        self._data = scipy.sparse.load_npz(DATA_PATH + self._data_filename + ".npz")
+        self.adj = np.load(DATA_PATH + self._data_filename + "_A.npy")
+        self.n_samples, self.modes = self._data.shape
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._count < self.n_samples:
+            self._count += 1
+            return self.__getitem__(self._count - 1)
+        self._count = 0
+        raise StopIteration
+
+    def _elem(self, i):
+        """Access the i-th element of the sparse array and output as a list."""
+        return list(self._data[i].toarray()[0])
+
+    def __getitem__(self, key):
+        if isinstance(key, (slice, tuple)):
+            if isinstance(key, tuple):
+                key = slice(*key)
+            range_tuple = key.indices(self.n_samples)
+            return [self._elem(i) for i in range(*range_tuple)]
+
+        if key < 0:
+            key += self.n_samples
+
+        return self._elem(key)

```

Comment:
Don't forget to test all these dunder methods thoroughly :laughing: Looks like there could be a lot of edge cases!

Reply:
I don't understand dunders :sob: :sob: :sob: